File: rasc_unl_flutter_app/lib/app/modules/competition/domain/models/competence_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'competence_model.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake, explicitToJson: true)
class CompetenceModel {
  final String id;
  final String name;
  final DateTime? competitionDate;
  final bool isActive;
  final bool isFinished; // Indica si la competencia finalizó
  final String createdBy; // User DNI of the creator
  final DateTime createdAt;
  final DateTime? updatedAt;

  // Sync fields
  final String syncStatus;
  final DateTime? lastSyncAt;
  final int version;
  final String? deviceId;
  final bool isDeleted;

  CompetenceModel({
    required this.id,
    required this.name,
    required this.competitionDate,
    required this.isActive,
    required this.isFinished,
    required this.createdBy,
    required this.createdAt,
    this.updatedAt,
    this.syncStatus = 'pending',
    this.lastSyncAt,
    this.version = 1,
    this.deviceId,
    this.isDeleted = false,
  });

  CompetenceModel copyWith({
    String? id,
    String? name,
    DateTime? competitionDate,
    bool? isActive,
    bool? isFinished,
    String? createdBy,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? syncStatus,
    DateTime? lastSyncAt,
    int? version,
    String? deviceId,
    bool? isDeleted,
  }) {
    return CompetenceModel(
      id: id ?? this.id,
      name: name ?? this.name,
      competitionDate: competitionDate ?? this.competitionDate,
      isActive: isActive ?? this.isActive,
      isFinished: isFinished ?? this.isFinished,
      createdBy: createdBy ?? this.createdBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      syncStatus: syncStatus ?? this.syncStatus,
      lastSyncAt: lastSyncAt ?? this.lastSyncAt,
      version: version ?? this.version,
      deviceId: deviceId ?? this.deviceId,
      isDeleted: isDeleted ?? this.isDeleted,
    );
  }

  factory CompetenceModel.fromJson(Map<String, dynamic> json) =>
      _$CompetenceModelFromJson(json);

  Map<String, dynamic> toJson() => _$CompetenceModelToJson(this);
}

File: rasc_unl_fast_api/app/modules/competencias/domain/models/competence_model.py
from sqlalchemy import Column, Integer, String, DateTime, Boolean, Date, JSON, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.db.database import Base
import enum
import uuid

class SyncStatus(str, enum.Enum):
    SYNCED = "synced"
    PENDING = "pending"
    CONFLICT = "conflict"

class CompetenceModel(Base):
    """
    Modelo de Competence
    Representa una competencia deportiva en el sistema.
    Relación 1:N con CompetitionRegistrationModel
    """
    __tablename__ = "competence"

    id = Column(String(36), primary_key=True, index=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(255), nullable=False, index=True)
    competition_date = Column(DateTime(timezone=True), nullable=False)  # Fecha y hora de inicio de competencia
    is_active = Column(Boolean, default=True, nullable=False)
    is_finished = Column(Boolean, default=False, nullable=False, server_default='false')  # Indica si la competencia finalizó
    
    created_by = Column(String(50), nullable=False)  # UserModel.dni
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Campos de sincronización
    sync_status = Column(Enum(SyncStatus), default=SyncStatus.PENDING, nullable=False)
    last_sync_at = Column(DateTime(timezone=True), nullable=True)
    version = Column(Integer, default=1, nullable=False)
    device_id = Column(String(255), nullable=True)
    is_deleted = Column(Boolean, default=False, nullable=False)

    # Relación 1:N con CompetitionRegistration
    competition_registration = relationship("CompetitionRegistrationModel", back_populates="competence", cascade="all, delete-orphan")
    
    
    def __repr__(self):
        return f"<Competence(id={self.id}, name='{self.name}')>"
